Delegate detection of the database version to the Hibernate ORM dialect and enable checks for Hibernate Reactive

yrodiere opened the pull request:
Fixes #43703
Fixes #42255
Fixes #45263

~~Based on #41310, which must be merged first. Critically, #41310 requires Hibernate ORM 7.0.0.Final to be released, so we won't be able to merge this PR for a while. I'm submitting it anyway as a draft, so that we remember it's that easy: only the last two commits will be needed after we rebase.~~ => #41310 was merged

yrodiere committed:
Delegate detection of the database version to the Hibernate ORM dialect

Because some dialects have very specific ways of detecting the version,
such as running some SQL, that work better than asking the JDBC driver.

yrodiere committed:
Re-enable detection of the database version by default when using explicit Hibernate ORM dialects

Because that detection is now done by the Hibernate ORM dialect itself,
and should work properly for all dialects.

DavideD reviewed (dismissed):
Before merging this, we should check that it works with Hibernate Reactive.

yrodiere committed:
Support DB version checks (enabled by default) with Hibernate Reactive

yrodiere commented:
I rebased on the ORM 7 upgrade PR (which upgrades to Reactive 3.0.0.CR1 too, now), and added tests to check DB version checking works with Hibernate Reactive, but so far it does not. I think `QuarkusRuntimeInitDialectFactory` is never used in reactive, leading to the DB version being unavailable and thus never checked.

I'll get back to this later, will try to fix it, and will update the Hibernate Reactive extension docs (since we need to document the DB version stuff there, too).

yrodiere commented:
@DavideD FYI, I updated the PR to _also_ perform version checks with Hibernate Reactive. It required a few changes, but it works! Thank you.

yrodiere committed:
Add non-regression tests for offline startup

yrodiere commented:
> There's definitely something wrong, and fortunately my new `DbOfflineOnStartupTest` seems to reproduce that. I'll need to have a closer look, but we'll miss the Quarkus 3.24.0.CR1 release. Well, we'll catch the next one :)

Should be good now.

yrodiere commented:
I added entries to the migration guide:

https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.25#database-versions-checks
https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.25#database-versions-checks-1
